python/torch_mlir/extras/fx_importer.py

```diff
@@ -1485,6 +1585,74 @@ def _import_torch_op_overload(
         for i, value in enumerate(operation.results):
             self.bind_node_value(node, value, i)
 
+    def _import_shape_symbols_with_guards(
+        self, loc: Location, symbolic_guards: Dict[str, RangeConstraint]
+    ):
+        for symbol, constraints in symbolic_guards.items():
+            # Create torch.sym_int ops
+            operation = Operation.create(
+                name="torch.symbolic_int",
+                attributes={
+                    "symbol_name": StringAttr.get(symbol),
+                    "min_val": self._cc.integer_attr(constraints.min_val, 64),
+                    "max_val": self._cc.integer_attr(constraints.max_val, 64),
+                },
+                results=[self._cc.torch_int_type],
+                loc=loc,
+            )
+            self.bind_symbol_value(symbol, operation.result)
+
+    def _create_bind_symbolic_shape_ops(self, loc: Location, node: torch_fx.Node):
```

Comment:
So if I understand correctly what this function is doing, it seems like the core of it is doing a kind of abstract evaluation by substituting the SymInt terminals with a corresponding AffineSymbolExpr, and then evaluating the expression tree as Python with the substitution. This then relies on AffineExpr having operator overloads and enough correspondence to be able to create an equivalent AffineExpr from the sympy expression tree. Do I have that right?

A couple thing I would wish for (assuming I have that right):

* Could we separate out the core of the abstract evaluation/expression transform into a dedicated top level helper function in the module. 
* This is the kind of thing that tends to have a number of corners to it that are hard to cover with e2e tests, and I would like to see a classic Python unit test that attempts to cover the weird cases, exercising the sympy_to_affine_expr_transform() helper.
* I'm not sure I trust the `affine_exprs = []` list comprehension yet. Does that only work for expression nodes that map 1:1 with Python operators and are overloaded on the AffineExpr?

There are some docs here https://docs.sympy.org/latest/tutorials/intro-tutorial/manipulation.html on how to do advanced expression manipulation and re-interpretation (example showing how to convert to latex, for example). I'd need to fiddle with it, but I would expect more of a walk that used a table of supported expressions that mapped explicitly to AffineExpr combiners, versus relying just on the accidental convergence of the two when evaluated as Python.

Reply:
Good callout. I'm revamping this to actually recursively traverse the sympy AST and translate it to AffineExpr depending on the type of sympy expr (Symbol vs Add vs Pow etc). Encountered a case of s0**2 (sympy.Pow) which AffineExpr don't support natively (i.e. s0**2 isn't but s0 * s0 is supported). I think most such cases should be handled with the mul, floordiv , ceildiv and mod from semi-affine maps : https://mlir.llvm.org/docs/Dialects/Affine/#semi-affine-maps, but as you said, needs a more systematic transation with the `sympy_to_affine_expr_transform` helper. Let me rework this part and get back to you.

Reply:
Took a first stab at `sympy_expr_to_semi_affine_expr()`. I've added the cases I could think of at the moment as being valid shape expressions (includes Add, Sub, Mul, Pow, Mod besides ConstantExpr / SymbolExpr). Needless to say, this will need to be massaged over time as we encounter fancier expressions that don't fit these categories. But glad that this seems much more extensible than its predecessor (python evaluations and operator overloads). 

I think this should address bullets 1 and 3 above (LMK if not). Bullet 2 is a reasonable ask - I will add python lit tests that unit test this function on some sympy expressions in a standalone manner, shortly.

Reply:
Added the python unit test for validating the `sympy_expr_to_semi_affine_expr` helper, addressing bullet 2 above.